Improve notifications view

Nitwel opened the pull request:
This PR is directly connected to the deprecation warning system that will be introduced in the following PR: #19508

Improves the notifications tab by the following:
- Being able to see the notification message (LOL)
- Not instantly redirecting you when clicking on a notification
- Adding pagination, searching, filtering
- Archive all button
- Cleaning up the code

New notifications tab.
![grafik](https://github.com/directus/directus/assets/22577866/37087316-838f-499f-9899-ad24660555c7)

Gif on how you interact with it.
![GIF 23 08 2023 14-28-34](https://github.com/directus/directus/assets/22577866/194a4d52-6ccb-43da-8a56-766dd3f34d61)

Nitwel committed:
improve notifications view

Nitwel committed:
add pagination

DanielBiegler commented:
> Being able to see the notification message (LOL)

lmao

Looks nice! We might wanna look into the console warnings?

> Invalid prop: type check failed for prop "clickable". Expected Boolean, got String with value "
> Hello Admin User,
> Admin User has mentioned you in a comment:
> ... etc ...

Nitwel committed:
add dense option

Nitwel committed:
cas to boolean

Nitwel committed:
add archive all button

Nitwel commented:
I've now also added an `archive all` button.
![grafik](https://github.com/directus/directus/assets/22577866/934d17e1-d2f6-420a-b780-f2dc052beec4)

Nitwel committed:
fix padding

DanielBiegler commented on app/src/lang/translations/en-US.yaml:
```diff
@@ -347,6 +347,7 @@ clear_items: Clear Items
 reset_to_default: Reset to Default
 undo_changes: Undo Changes
 notifications: Notifications
+mark_all_as_read: Archive All
```
I do think `read` and `archive` are different enough in meaning that we should change the key. Agree?

```suggestion
archive_all: Archive All
```

Maybe? Or something like

```suggestion
archive_all_items: Archive All
```

Nitwel committed:
make links colored

Nitwel committed:
Update app/src/lang/translations/en-US.yaml

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

Nitwel committed:
Update notifications-drawer.vue

DanielBiegler reviewed (dismissed):
Yay, lovely!

azrikahar commented on app/src/views/private/components/notifications-drawer.vue:
```diff
@@ -39,34 +42,78 @@
+							:model-value="selection.includes(notification.id)"
+							@update:model-value="toggleSelected(notification.id)"
+						/>
+						<div v-tooltip="notification.subject" class="title">{{ notification.subject }}</div>
+						<div class="spacer" />
+						<div class="time">{{ notification.time }}</div>
```
Thoughts on using `v-text-overflow` component for the title and the `datetime` display for the time? It's working perfectly as it is now, but just wondering should we reuse components more (very pedantic question!)

paescuj replied:
Yes, we should 👍 
I've updated it to use `v-text-overflow`. The `datetime` display doesn't come with a tooltip when there's an overflow, though. So I think we should move the logic from `datetime` into a composable which we can then reuse and pass to `v-text-overflow` here.

azrikahar reviewed (commented):
"Improve" is definitely an understatement here, love it a lot!

Just a tiny question regarding the `Archive All` action button added/mentioned [over here](https://github.com/directus/directus/pull/19503#issuecomment-1689976899), it feels like a text action button slightly uncommon in terms of action buttons in general that are rounded icons (I could be very wrong here), so just tapping in @rijkvanzanten for some thoughts on it.

br41nslug reviewed (dismissed):
Attempted to do some xss but unfortunately it seems to get sanitized properly 😂 Great improvement

paescuj committed:
Use v-text-overflow component for title & time, clean-up styles

paescuj committed:
Load notifications only as soon as drawer is opened

paescuj committed:
Fix link when item is undefined

paescuj commented:
> "Improve" is definitely an understatement here, love it a lot!

Yeah 🚀 

> Just a tiny question regarding the `Archive All` action button added/mentioned [over here](https://github.com/directus/directus/pull/19503#issuecomment-1689976899), it feels like a text action button slightly uncommon in terms of action buttons in general that are rounded icons (I could be very wrong here), so just tapping in @rijkvanzanten for some thoughts on it.

I agree! Overall, the design feels a bit exotic in comparison to the rest of the app. Let's replace the "Archive All" button, I think the rest is fine for now 👍 Did a quick test and I feel like the header bar from the card layout would be a good match here (may additionally serve for sorting):

<img src="https://github.com/directus/directus/assets/5363448/d31957ea-9076-40f7-8483-9f3be38dc244">

Nitwel commented:
The reason for the text based button was that I didn't find an icon that would explain the difference between the archive and archive all button. And a select all button is not an option because we're using pagination so we don't know what id's there are on the next page. Best solution to me would be to find a proper icon for the archieve all button.

paescuj commented:
> The reason for the text based button was that I didn't find an icon that would explain the difference between the archive and archive all button. And a select all button is not an option because we're using pagination so we don't know what id's there are on the next page.

I see, but it might be enough to be able to select all items on the current page, like in the other layouts 🤔 

>Best solution to me would be to find a proper icon for the archieve all button. @benhaynes thoughts for an icon?
> But I really like the idea of adding search and filtering to notifications so will add that now too.

Yeah, I suggest waiting for @benhaynes and/or @rijkvanzanten input before proceeding with updates to the UI 👍

Nitwel commented:
Gave it some finishing touches. You are now able to search and filer. I simplified the archieve all button to an icon that should be fitting. And it now reuses as much code as possible.

![grafik](https://github.com/directus/directus/assets/22577866/e7c9b93b-6399-4b96-a3ce-1bdd5d4893ce)
![grafik](https://github.com/directus/directus/assets/22577866/f81df3c2-66dc-4fe4-b208-dcef0645130c)

Nitwel committed:
fix resetting page

paescuj committed:
Start search bar in collapsed state when opening drawer

paescuj committed:
Effectively apply smaller width to search bar

paescuj committed:
Custom sidebar label

paescuj committed:
Remove unused/obsolete `useTimeFromNow` composable

paescuj committed:
Move datetime logic to reusable composable & use it in notification drawer

Nice benefit is that it now uses the same refresh interval for all usages of `useDatetime` across the whole page

paescuj committed:
Update unread count on archive actions

benhaynes commented:
That's the right icon! Love it...

DanielBiegler commented:
We should probably consider merging this into our new pre-release-branch instead of main for the bigger release? Agree?

paescuj committed:
Finalize

- Transform use-datetime composable into renderless component,
  see https://vuejs.org/guide/components/slots.html#scoped-slots

  With this we can get rid of the hacky 'refresher' but still use it in
  any context, e.g. together with v-text-overflow

- Only update unread count on archive all action as we cannot be sure
  about the current count in case of single archive/unarchive

- Renaming some variables/classes for clarity

paescuj committed:
Merge branch 'main' into fix-notification

paescuj committed:
Add missing newline

paescuj committed:
Add changeset

paescuj reviewed (approved):
Nice work @Nitwel! I like the new UI ❤️

paescuj commented:
> We should probably consider merging this into our new pre-release-branch instead of main for the bigger release? Agree?

I think this should be fine for `main` now 👍 

The amount of actual changes has been reduced in https://github.com/directus/directus/pull/19503/commits/b451b83c932d27d650a39a50b5237abc507bb42c.

br41nslug reviewed (changes requested):
Currently the notifications have the CSS style `user-select: none` applied preventing you from selecting the text in the body of the notification. I think this text should be selectable as it may contain information you'd like to copy-paste.

Also this signout button really needs to go, i clicked it an insane amount of times while testing this PR but that doesnt have to be done in this PR.

https://github.com/directus/directus/assets/9389634/646655cc-309d-4eab-ba6e-30a4cd1983a8

Besides that looking very good 😄

paescuj committed:
Fix optional props definition

paescuj committed:
Make notification content selectable

paescuj committed:
Reuse markdown style from markdown interface

paescuj committed:
Generalize small search input if sidebar active
